p2p/database/datastore.go

```diff
@@ -127,6 +129,76 @@ type DatastoreOptions struct {
 	TTL                          time.Duration
 }
 
+const (
+	shardCount      = 32
+	cleanupInterval = 10 * time.Minute // Interval between cleanups
+	entryTTL        = 1 * time.Hour    // TTL for each entry
+)
+
+var (
+	dedupMaps [shardCount]map[string]time.Time
+	mutexes   [shardCount]sync.Mutex
+)
+
+// Initialize dedupMaps
+func init() {
+	// Initialize each shard in dedupMaps
+	for i := range dedupMaps {
+		dedupMaps[i] = make(map[string]time.Time)
+	}
+	// Start the cleanup routine to periodically remove stale entries
+	StartCleanupRoutine()
+}
+
+// getShard retrieves the shard and mutex for a given key
+func getShard(key string) (map[string]time.Time, *sync.Mutex) {
+	hash := fnv.New32a()
+	hash.Write([]byte(key))
+	shardIndex := hash.Sum32() % shardCount
+	return dedupMaps[shardIndex], &mutexes[shardIndex]
+}
+
+// Deduplication function that checks for uniqueness and writes if unique
+func (d *Datastore) DeduplicateAndWriteEvent(peer *enode.Node, eventKind string, hash common.Hash, hashKind string, tfs time.Time) {
+	key := fmt.Sprintf("%s_%s_%s", d.sensorID, peer.URLv4(), hash.Hex())
+	shard, mu := getShard(key)
+
+	mu.Lock()
+	defer mu.Unlock()
+
+	if existingTime, found := shard[key]; found && !tfs.Before(existingTime) {
+		return // Skip if not the earliest
+	}
+	shard[key] = tfs
+	d.writeEvent(peer, eventKind, hash, hashKind, tfs)
+}
+
+// Cleanup function that removes expired entries from each shard
+func cleanupDedupMap() {
+	for {
+		time.Sleep(cleanupInterval) // Wait for the next cleanup cycle
+
+		expiryTime := time.Now().Add(-entryTTL)
+		for i := 0; i < shardCount; i++ {
+			mu := &mutexes[i]
+			shard := dedupMaps[i]
+
+			mu.Lock()
+			for key, timestamp := range shard {
+				if timestamp.Before(expiryTime) {
+					delete(shard, key) // Remove stale entry
```

Comment:
The thing I encountered before is that `delete` on maps doesn't actually decrease the size of the map (but it does clear the key). It retains the maximum size that it was, which caused some memory issues in earlier sensor iteraitons.

Reply:
interesting, thanks for pointing out, will reexamine

Reply:
adjusted per your feedback